### PR TITLE
Add sync/async mode to send data

### PR DESCRIPTION
**Pull request by tuteng**
### Motivation

Add options for message

### Modifications
* add sync/async mode to send data, default sync
* add more options for message

### Result
Use the following configuration file to test headers and syncMode：
```
# example.conf: A single-node Flume configuration

# Name the components on this agent
a1.sources = r1
a1.sinks = k1
a1.channels = c1

# Describe/configure the source
a1.sources.r1.type = spooldir
a1.sources.r1.spoolDir = /logs
a1.sources.r1.fileHeader = true
#a1.sources.r1.bind = 0.0.0.0
#a1.sources.r1.port = 44445


## Describe the sink
a1.sinks.k1.type = org.apache.flume.sink.pulsar.PulsarSink
a1.sinks.k1.serviceUrl = pulsar-flume-standalone:6650
a1.sinks.k1.topicName = test
a1.sinks.k1.producerName = testProducer
a1.sinks.k1.syncMode = false

# Use a channel which buffers events in memory
a1.channels.c1.type = memory
a1.channels.c1.capacity = 1000
a1.channels.c1.transactionCapacity = 1000

# Bind the source and sink to the channel
a1.sources.r1.channels = c1
a1.sinks.k1.channel = c1
```

### CODE REVIEW

File: src/main/java/org/apache/flume/sink/pulsar/PulsarSink.java [hunk 1]
```diff
@@ -203,11 +208,28 @@ public Status process() throws EventDeliveryException {
...
+                }
+                if (syncMode) {
+                    newMessage.send();
+                } else {
+                    newMessage.sendAsync().thenAccept(msgId -> {
+                        log.info("Success send message in asyncMode {}", msgId);
+                    });
+                }
+            }
+            if (syncMode) {
```
Comment (sijie) [inline src/main/java/org/apache/flume/sink/pulsar/PulsarSink.java]: @tuteng if `syncMode` is enabled, you don't need to flush. `send()` only returns when a message is flushed and persistent.

I don't think `flushAsync` should be used here. based on my understanding on flume transaction, you have to wait until all messages are flushed and persisted before committing the transaction.

File: src/main/java/org/apache/flume/sink/pulsar/PulsarSink.java [hunk 2]
```diff
@@ -203,11 +208,28 @@ public Status process() throws EventDeliveryException {
...
+                    if (event.getHeaders().get("key") != null) {
+                        newMessage = newMessage.key(event.getHeaders().get("key"));
+                    }
+                    newMessage.value(serializeEvent(event, useAvroEventFormat)).properties(event.getHeaders());
+                }
+                if (syncMode) {
+                    newMessage.send();
+                } else {
+                    newMessage.sendAsync().thenAccept(msgId -> {
+                        log.info("Success send message in asyncMode {}", msgId);
```
Comment (sijie) [inline src/main/java/org/apache/flume/sink/pulsar/PulsarSink.java]: hmm. this log will be printed every message, no?

Comment (tuteng) [reply src/main/java/org/apache/flume/sink/pulsar/PulsarSink.java]: Sorry, i forgot it

File: src/main/java/org/apache/flume/sink/pulsar/PulsarSink.java (hunk 1 above)
Comment (tuteng) [reply src/main/java/org/apache/flume/sink/pulsar/PulsarSink.java]: good, i see

File: src/main/java/org/apache/flume/sink/pulsar/PulsarSink.java [hunk 3]
```diff
@@ -203,11 +208,21 @@ public Status process() throws EventDeliveryException {
...
+                    }
+                    newMessage.value(serializeEvent(event, useAvroEventFormat)).properties(event.getHeaders());
+                }
+                if (syncMode) {
+                    newMessage.send();
+                } else {
+                    newMessage.sendAsync();
+                }
             }
             transaction.commit();
```
Comment (sijie) [inline src/main/java/org/apache/flume/sink/pulsar/PulsarSink.java]: we should add following logic before commit.

```
if (!syncMode) {
    producer.flush();
}
```

**Pull request merged**